File: plugins/agent-skills/skills/claude-delegate/scripts/claude_delegate/workspace.py
from __future__ import annotations

import hashlib
from pathlib import Path

from .execution_workspace import ExecutionWorkspace


# Maximum file size (in bytes) to include content in snapshot for patch generation
_MAX_CONTENT_SIZE = 1024 * 1024  # 1 MB


def _fingerprint(path: Path) -> dict:
    hasher = hashlib.sha1()
    with path.open("rb") as handle:
        while True:
            chunk = handle.read(1024 * 1024)
            if not chunk:
                break
            hasher.update(chunk)
    stat = path.stat()
    return {
        "mtime_ns": stat.st_mtime_ns,
        "sha1": hasher.hexdigest(),
        "size": stat.st_size,
    }


def _capture_content(path: Path) -> str | None:
    """Capture file content if under size limit. Returns None for binary or too-large files."""
    stat = path.stat()
    if stat.st_size > _MAX_CONTENT_SIZE:
        return None
    try:
        return path.read_text(encoding="utf-8", errors="strict")
    except (UnicodeDecodeError, OSError):
        return None


def has_semantic_change(before_entry: dict | None, after_entry: dict | None) -> bool:
    """Return whether file existence or content changed, ignoring mtime-only churn."""
    if (before_entry is None) != (after_entry is None):
        return True
    if before_entry is None:
        return False
    return (before_entry.get("sha1"), before_entry.get("size")) != (
        after_entry.get("sha1"),
        after_entry.get("size"),
    )


def _should_exclude_path(resolved: Path, excludes: list[Path], exclude_globs: list[str]) -> bool:
    if any(resolved == excluded or excluded in resolved.parents for excluded in excludes):
        return True
    for pattern in exclude_globs:
        if resolved.match(pattern):
            return True
    return False


def capture_workspace_state(
    *,
    workspace: ExecutionWorkspace,
    observe_roots: list[str],
    exclude_roots: list[str],
    exclude_globs: list[str],
) -> dict[str, dict]:
    state: dict[str, dict] = {}
    excludes = [Path(item).resolve() for item in exclude_roots]
    for raw_root in observe_roots:
        source_root = workspace.resolve_source_path(raw_root)
        execution_root = workspace.map_source_to_execution(source_root)
        if not execution_root.exists():
            continue
        candidates = [execution_root] if execution_root.is_file() else sorted(path for path in execution_root.rglob("*") if path.is_file())
        for execution_path in candidates:
            resolved = execution_path.resolve()
            if _should_exclude_path(resolved, excludes, exclude_globs):
                continue
            source_path = workspace.map_execution_to_source(resolved)
            entry = _fingerprint(resolved)
            entry["execution_path"] = str(resolved)
            entry["source_path"] = str(source_path)
            # Include content for patch generation (up to size limit)
            content = _capture_content(resolved)
            if content is not None:
                entry["content"] = content
            state[str(source_path)] = entry
    return state


def diff_workspace_state(before: dict[str, dict], after: dict[str, dict], source_cwd: str) -> list[dict]:
    changes: list[dict] = []
    all_paths = sorted(set(before) | set(after))
    cwd_path = Path(source_cwd).resolve()

    for path_str in all_paths:
        before_entry = before.get(path_str)
        after_entry = after.get(path_str)
        if not has_semantic_change(before_entry, after_entry):
            continue
        path = Path(path_str)
        try:
            relative = str(path.relative_to(cwd_path))
        except ValueError:
            relative = path_str

        if before_entry is None:
            change = "added"
        elif after_entry is None:
            change = "deleted"
        else:
            change = "modified"

        execution_path = (after_entry or before_entry or {}).get("execution_path")
        changes.append(
            {
                "change": change,
                "path": path_str,
                "execution_path": execution_path,
                "relative_path": relative,
            }
        )

    return changes

File: plugins/agent-skills/skills/claude-delegate/src/lib/workspace.ts
import { createHash } from "node:crypto";
import {
  readFileSync,
  readdirSync,
  statSync,
} from "node:fs";
import path from "node:path";

import type { ExecutionWorkspace } from "./execution-workspace.js";
import {
  mapExecutionToSource,
  mapSourceToExecution,
  resolveSourcePath,
} from "./execution-workspace.js";

/** Maximum file size (bytes) to include content in snapshot for patch generation. */
const MAX_CONTENT_SIZE = 1024 * 1024; // 1 MB

function fingerprint(filePath: string): Record<string, unknown> {
  const hasher = createHash("sha1");
  const data = readFileSync(filePath);
  hasher.update(data);
  const stat = statSync(filePath);
  return {
    mtime_ns: Math.round(stat.mtimeMs * 1e6),
    sha1: hasher.digest("hex"),
    size: stat.size,
  };
}

function captureContent(filePath: string): string | null {
  const stat = statSync(filePath);
  if (stat.size > MAX_CONTENT_SIZE) return null;
  try {
    const buf = readFileSync(filePath);
    // Reject non-UTF-8: decode with throw option equivalent
    const text = new TextDecoder("utf-8", { fatal: true }).decode(buf);
    return text;
  } catch {
    return null;
  }
}

export function hasSemanticChange(
  beforeEntry: Record<string, unknown> | null,
  afterEntry: Record<string, unknown> | null,
): boolean {
  if ((beforeEntry === null) !== (afterEntry === null)) return true;
  if (beforeEntry === null) return false;
  return (
    beforeEntry.sha1 !== afterEntry!.sha1 ||
    beforeEntry.size !== afterEntry!.size
  );
}

function shouldExcludePath(
  resolved: string,
  excludes: string[],
  excludeGlobs: string[],
): boolean {
  for (const excluded of excludes) {
    if (resolved === excluded || resolved.startsWith(excluded + path.sep)) {
      return true;
    }
  }
  for (const pattern of excludeGlobs) {
    // path.basename match for simple glob patterns (e.g. "*.pyc", "node_modules")
    const basename = path.basename(resolved);
    if (simpleGlobMatch(basename, pattern) || simpleGlobMatch(resolved, pattern)) {
      return true;
    }
  }
  return false;
}

/** Minimal glob matching: supports * as wildcard for simple patterns. */
function simpleGlobMatch(text: string, pattern: string): boolean {
  // Convert glob to regex: escape dots, replace * with .*
  const regexStr =
    "^" +
    pattern
      .replace(/[.+^${}()|[\]\\]/g, "\\$&")
      .replace(/\*\*/g, "\0")
      .replace(/\*/g, "[^/]*")
      .replace(/\0/g, ".*") +
    "$";
  return new RegExp(regexStr).test(text);
}

/** Recursively walk a directory, returning sorted file paths. */
function walkFiles(root: string): string[] {
  const result: string[] = [];
  const entries = readdirSync(root, { withFileTypes: true });
  for (const entry of entries) {
    const full = path.join(root, entry.name);
    if (entry.isDirectory()) {
      result.push(...walkFiles(full));
    } else if (entry.isFile()) {
      result.push(full);
    }
  }
  return result.sort();
}

function existsSync(p: string): boolean {
  try {
    statSync(p);
    return true;
  } catch {
    return false;
  }
}

function isFile(p: string): boolean {
  try {
    return statSync(p).isFile();
  } catch {
    return false;
  }
}

export function captureWorkspaceState(opts: {
  workspace: ExecutionWorkspace;
  observeRoots: string[];
  excludeRoots: string[];
  excludeGlobs: string[];
}): Record<string, Record<string, unknown>> {
  const state: Record<string, Record<string, unknown>> = {};
  const excludes = opts.excludeRoots.map((item) => path.resolve(item));

  for (const rawRoot of opts.observeRoots) {
    const sourceRoot = resolveSourcePath(opts.workspace, rawRoot);
    const executionRoot = mapSourceToExecution(opts.workspace, sourceRoot);
    if (!existsSync(executionRoot)) continue;

    const candidates = isFile(executionRoot)
      ? [executionRoot]
      : walkFiles(executionRoot);

    for (const executionPath of candidates) {
      const resolved = path.resolve(executionPath);
      if (shouldExcludePath(resolved, excludes, opts.excludeGlobs)) continue;

      const sourcePath = mapExecutionToSource(opts.workspace, resolved);
      const entry = fingerprint(resolved);
      entry.execution_path = resolved;
      entry.source_path = sourcePath;

      const content = captureContent(resolved);
      if (content !== null) {
        entry.content = content;
      }
      state[sourcePath] = entry;
    }
  }
  return state;
}

export function diffWorkspaceState(
  before: Record<string, Record<string, unknown>>,
  after: Record<string, Record<string, unknown>>,
  sourceCwd: string,
): Record<string, unknown>[] {
  const changes: Record<string, unknown>[] = [];
  const allPaths = [...new Set([...Object.keys(before), ...Object.keys(after)])].sort();
  const cwdPath = path.resolve(sourceCwd);

  for (const pathStr of allPaths) {
    const beforeEntry = before[pathStr] ?? null;
    const afterEntry = after[pathStr] ?? null;
    if (!hasSemanticChange(beforeEntry, afterEntry)) continue;

    let relative: string;
    const rel = path.relative(cwdPath, pathStr);
    relative = rel.startsWith("..") ? pathStr : rel;

    let change: string;
    if (beforeEntry === null) {
      change = "added";
    } else if (afterEntry === null) {
      change = "deleted";
    } else {
      change = "modified";
    }

    const executionPath =
      (afterEntry ?? beforeEntry ?? {}).execution_path ?? null;
    changes.push({
      change,
      path: pathStr,
      execution_path: executionPath,
      relative_path: relative,
    });
  }

  return changes;
}
